presto-main/src/main/java/com/facebook/presto/type/khyperloglog/KHyperLogLog.java

```diff
@@ -0,0 +1,351 @@
+        hll.add(uii);
+        increaseTotalHllSize(hll);
+
+        removeOverflowEntries();
+    }
+
+    public long cardinality()
+    {
+        if (isExact()) {
+            return minhash.size();
+        }
+
+        // Intuition is: get the stored hashes' density, and extrapolate to the whole Hash output range.
+        // Since Hash output range (2^64) cannot be stored in long type, I use half of the range
+        // via Long.MAX_VALUE and also divide the hash values' density by 2. The "-1" is bias correction
+        // detailed in "On Synopses for Distinct-Value Estimation Under Multiset Operations" by Beyer et. al.
+        long hashesRange = minhash.lastLongKey() - Long.MIN_VALUE;
+        double halfDensity = Long.divideUnsigned(hashesRange, minhash.size()) / 2D;
+        return (long) (HASH_OUTPUT_HALF_RANGE / halfDensity);
+    }
+
+    public KHyperLogLog mergeWith(KHyperLogLog other)
+    {
+        LongIterator iterator = other.minhash.keySet().iterator();
+        while (iterator.hasNext()) {
+            long key = iterator.nextLong();
+            HyperLogLog thisHll = minhash.get(key);
+            HyperLogLog otherHll = other.minhash.get(key);
+            if (minhash.containsKey(key)) {
+                decreaseTotalHllSize(thisHll);
+                thisHll.mergeWith(otherHll);
+                increaseTotalHllSize(thisHll);
+            }
+            else {
+                minhash.put(key, otherHll);
+                increaseTotalHllSize(otherHll);
+            }
+        }
+
+        removeOverflowEntries();
+
+        return this;
+    }
+
+    public double reidentificationPotential(long threshold)
+    {
+        int highlyUniqueValues = 0;
+        for (HyperLogLog hll : minhash.values()) {
+            if (hll.cardinality() <= threshold) {
+                highlyUniqueValues++;
+            }
+        }
+        return (double) highlyUniqueValues / minhash.size();
+    }
+
+    public Long2DoubleMap uniquenessDistribution()
+    {
+        return uniquenessDistribution(minhash.size());
+    }
+
+    public Long2DoubleMap uniquenessDistribution(long histogramSize)
+    {
+        Long2DoubleMap out = new Long2DoubleOpenHashMap();
+        PrimitiveIterator.OfLong iterator = LongStream.rangeClosed(1, histogramSize).iterator();
+        while (iterator.hasNext()) {
+            // Initialize all entries to zero
+            out.put(iterator.nextLong(), 0D);
+        }
+
+        int size = minhash.size();
+        for (HyperLogLog hll : minhash.values()) {
+            long bucket = Math.min(hll.cardinality(), histogramSize);
+            out.merge(bucket, (double) 1 / size, Double::sum);
+        }
+        return out;
+    }
+
+    private void removeOverflowEntries()
+    {
+        while (minhash.size() > maxSize) {
+            HyperLogLog hll = minhash.remove(minhash.lastLongKey());
+            decreaseTotalHllSize(hll);
+        }
+    }
+
+    private void decreaseTotalHllSize(HyperLogLog hll)
```

Comment:
You can just have one method which takes in the memory size in bytes.  We can pass in the negative value to subtract and positive value to add.

Reply:
But then we would need one method call for InMemory size and one method call for serialized size